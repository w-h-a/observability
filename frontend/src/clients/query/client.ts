export interface IClient {
	get<T = any>(url: string): Promise<{ data: T }>;
}
