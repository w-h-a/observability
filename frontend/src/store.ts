import { applyMiddleware, combineReducers, compose, createStore } from "redux";
import { thunk } from "redux-thunk";

export const reducers = combineReducers({});

export const store = createStore(reducers, compose(applyMiddleware(thunk)));
