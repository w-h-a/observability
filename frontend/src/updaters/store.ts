import { combineReducers, configureStore } from "@reduxjs/toolkit";
import { ServicesUpdater } from "./services/servicesTable";
import { TimeUpdater } from "./time/time";
import { ServiceUpdater } from "./service/service";

const reducer = combineReducers({
	maxMinTime: TimeUpdater.MaxMinTimeReducer,
	services: ServicesUpdater.ServicesReducer,
	endpoints: ServiceUpdater.EndpointsReducer,
});

export const store = configureStore({
	reducer: reducer,
});

export type RootState = ReturnType<typeof store.getState>;

export type AppDispatch = typeof store.dispatch;
