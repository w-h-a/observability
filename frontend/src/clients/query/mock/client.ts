import { IClient } from "../client";

export class Client implements IClient {
	async get<T = any>(url: string): Promise<{ data: T }> {
		return {
			data: [] as T,
		};
	}
}
