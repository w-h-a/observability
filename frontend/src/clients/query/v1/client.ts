import axios, { AxiosInstance } from "axios";
import { IClient } from "../client";

export class Client implements IClient {
	private client: AxiosInstance = axios.create();

	async get<T = any>(url: string): Promise<{ data: T }> {
		const result = await this.client.get<T>(url);
		return result;
	}
}
