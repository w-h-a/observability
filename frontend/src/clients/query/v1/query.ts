import { Config } from "../../../config/config";
import { IClient } from "../client";

export class Query {
	static async GetServices<T = any>(
		client: IClient,
		start: number,
		end: number,
	): Promise<{ data: T }> {
		try {
			const path = `/services`;

			const query = `?start=${start}&end=${end}`;

			return await client.get<T>(
				`${Config.GetInstance().get("baseUrl")}${path}${query}`,
			);
		} catch (err: unknown) {
			console.log(`query client failed to retrieve services: ${err}`);
			throw err;
		}
	}

	static async GetEndpoints<T = any>(
		client: IClient,
		start: number,
		end: number,
		serviceName: string,
	): Promise<{ data: T }> {
		try {
			const path = `/service/endpoints`;

			const query = `?start=${start}&end=${end}&service=${serviceName}`;

			return await client.get<T>(
				`${Config.GetInstance().get("baseUrl")}${path}${query}`,
			);
		} catch (err: unknown) {
			console.log(`query client failed to retrieve endpoints: ${err}`);
			throw err;
		}
	}
}
