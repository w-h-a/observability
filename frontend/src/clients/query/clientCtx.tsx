import { createContext } from "react";
import { Client } from "./v1/client";
import { IClient } from "./client";

export const ClientContext = createContext<{ queryClient: IClient }>({
	queryClient: new Client(),
});
