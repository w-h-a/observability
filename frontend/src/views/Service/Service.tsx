import { useContext, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useParams } from "react-router-dom";
import { Card, Col, Row, Tabs } from "antd";
import { EndpointsTable } from "./EndpointsTable";
import { RequestRateChart } from "./RequestRateChart";
import { AppDispatch, RootState } from "../../updaters/store";
import { ServiceUpdater } from "../../updaters/service/service";
import { ClientContext } from "../../clients/query/clientCtx";

export const Service = () => {
	const { queryClient } = useContext(ClientContext);

	const { service } = useParams<{ service: string }>();

	const maxMinTime = useSelector((state: RootState) => state.maxMinTime);
	const endpoints = useSelector((state: RootState) => state.endpoints);
	const serviceMetrics = useSelector((state: RootState) => state.serviceMetrics);

	const dispatch: AppDispatch = useDispatch();

	useEffect(() => {
		dispatch(ServiceUpdater.Endpoints(queryClient, maxMinTime, service));
		dispatch(ServiceUpdater.ServiceMetrics(queryClient, maxMinTime, service));
	}, [dispatch, queryClient, maxMinTime, service]);

	return (
		<Tabs defaultActiveKey="1">
			<Tabs.TabPane tab="Service Metrics" key="1">
				<Row gutter={32} style={{ margin: 20 }}>
					<Col span={12}>
						{/* <Card bodyStyle={{ padding: 10 }}>
							<LatencyLineChart
								data={serviceMetrics}
								// popupClickHandler={onTracePopupClick}
							/>
						</Card> */}
					</Col>
					<Col span={12}>
						<Card bodyStyle={{ padding: 10 }}>
							<RequestRateChart data={serviceMetrics} />
						</Card>
					</Col>
				</Row>
				<Row gutter={32} style={{ margin: 20 }}>
					<Col span={12}>
						{/* <Card bodyStyle={{ padding: 10 }}>
							<ErrorRateChart
								data={serviceMetrics}
								// popupClickHandler={onErrTracePopupClick}
							/>
						</Card> */}
					</Col>
					<Col span={12}>
						<EndpointsTable data={endpoints} />
					</Col>
				</Row>
			</Tabs.TabPane>
		</Tabs>
	);
};
