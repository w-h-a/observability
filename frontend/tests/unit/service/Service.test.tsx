import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { render, RenderResult } from "@testing-library/react";
import { TestCase } from "../testcase";
import { Service } from "../../../src/views/Service/Service";
import { store } from "../../../src/updaters/store";
import { IClient } from "../../../src/clients/query/client";
import { Client } from "../../../src/clients/query/mock/client";
import { ClientContext } from "../../../src/clients/query/clientCtx";

vi.mock("../../../src/updaters/time/utils", () => {
	return {
		Now: vi.fn(() => 1736393070 * 1000),
	};
});

const service = "frontend";

vi.mock("react-router-dom", async () => {
	const mod = await vi.importActual("react-router-dom");
	return {
		...mod,
		useParams: () => ({
			service: service,
		}),
	};
});

describe("Service", () => {
	const mockQueryClient: IClient = new Client();

	const testCases: TestCase[] = [
		{
			When:
				"when: the user goes to the service page and calls to the backend are a success",
			GetStub: function <T>(url: string): Promise<{ data: T }> {
				return new Promise((resolve) => {
					resolve({
						data: [
							{
								name: "/config",
								p50: 0.07 * 1000000,
								p95: 0.09 * 1000000,
								p99: 0.1 * 1000000,
								numCalls: 8,
							},
							{
								name: "/dispatch",
								p50: 765.25 * 1000000,
								p95: 904.79 * 1000000,
								p99: 937.2 * 1000000,
								numCalls: 6,
							},
						] as T,
					});
				});
			},
			Then: "then: we render the service-specific tabs and tables",
			ClientCalledTimes: 1,
			ClientCalledWith: [
				`http://localhost:4000/api/v1/service/endpoints?start=1736392170&end=1736393070&service=${service}`,
			],
		},
	];

	testCases.forEach((testCase) => {
		describe(testCase.When, () => {
			let view: RenderResult;

			beforeEach(async () => {
				mockQueryClient.get = vi.fn().mockImplementation(testCase.GetStub);

				view = render(
					<Provider store={store}>
						<BrowserRouter>
							<ClientContext.Provider value={{ queryClient: mockQueryClient }}>
								<Service />
							</ClientContext.Provider>
						</BrowserRouter>
					</Provider>,
				);
			});

			afterEach(() => {
				vi.clearAllMocks();
			});

			test(testCase.Then, async () => {
				expect(mockQueryClient.get).toHaveBeenCalledTimes(
					testCase.ClientCalledTimes,
				);

				for (let count = 1; count <= testCase.ClientCalledTimes; count++) {
					expect(mockQueryClient.get).toHaveBeenCalledWith(
						testCase.ClientCalledWith[count - 1],
					);
				}

				expect(view.container).toMatchSnapshot();
			});
		});
	});
});
