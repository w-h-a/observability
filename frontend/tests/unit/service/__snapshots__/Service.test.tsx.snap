// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Service > when: the user goes to the service page and calls to the backend are a success > then: we render the service-specific tabs and tables 1`] = `
<div>
  <div
    class="ant-tabs ant-tabs-top css-dev-only-do-not-override-7ny38l"
  >
    <div
      aria-orientation="horizontal"
      class="ant-tabs-nav"
      role="tablist"
    >
      <div
        class="ant-tabs-nav-wrap"
      >
        <div
          class="ant-tabs-nav-list"
          style="transform: translate(0px, 0px);"
        >
          <div
            class="ant-tabs-tab ant-tabs-tab-active"
            data-node-key="1"
          >
            <div
              aria-controls="rc-tabs-test-panel-1"
              aria-selected="true"
              class="ant-tabs-tab-btn"
              id="rc-tabs-test-tab-1"
              role="tab"
              tabindex="0"
            >
              Service Metrics
            </div>
          </div>
          <div
            class="ant-tabs-ink-bar ant-tabs-ink-bar-animated"
          />
        </div>
      </div>
      <div
        class="ant-tabs-nav-operations ant-tabs-nav-operations-hidden"
      >
        <button
          aria-controls="rc-tabs-test-more-popup"
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-hidden="true"
          class="ant-tabs-nav-more"
          id="rc-tabs-test-more"
          style="visibility: hidden; order: 1;"
          tabindex="-1"
          type="button"
        >
          <span
            aria-label="ellipsis"
            class="anticon anticon-ellipsis"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="ellipsis"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
    <div
      class="ant-tabs-content-holder"
    >
      <div
        class="ant-tabs-content ant-tabs-content-top"
      >
        <div
          aria-hidden="false"
          aria-labelledby="rc-tabs-test-tab-1"
          class="ant-tabs-tabpane ant-tabs-tabpane-active"
          id="rc-tabs-test-panel-1"
          role="tabpanel"
          tabindex="0"
        >
          <div
            class="ant-row css-dev-only-do-not-override-7ny38l"
            style="margin: 20px;"
          >
            <div
              class="ant-col ant-col-12 css-dev-only-do-not-override-7ny38l"
              style="padding-left: 16px; padding-right: 16px;"
            >
              <div
                class="ant-card ant-card-bordered css-dev-only-do-not-override-7ny38l"
              >
                <div
                  class="ant-card-body"
                  style="padding: 10px;"
                >
                  <div>
                    <div
                      style="text-align: center;"
                    >
                      Latency in ms
                    </div>
                    <canvas
                      class="chartjs-render-monitor"
                      height="0"
                      style="display: block;"
                      width="0"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="ant-col ant-col-12 css-dev-only-do-not-override-7ny38l"
              style="padding-left: 16px; padding-right: 16px;"
            >
              <div
                class="ant-card ant-card-bordered css-dev-only-do-not-override-7ny38l"
              >
                <div
                  class="ant-card-body"
                  style="padding: 10px;"
                >
                  <div>
                    <div
                      style="text-align: center;"
                    >
                      Request per sec
                    </div>
                    <canvas
                      class="chartjs-render-monitor"
                      height="0"
                      style="display: block;"
                      width="0"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ant-row css-dev-only-do-not-override-7ny38l"
            style="margin: 20px;"
          >
            <div
              class="ant-col ant-col-12 css-dev-only-do-not-override-7ny38l"
              style="padding-left: 16px; padding-right: 16px;"
            />
            <div
              class="ant-col ant-col-12 css-dev-only-do-not-override-7ny38l"
              style="padding-left: 16px; padding-right: 16px;"
            >
              <div>
                <h6>
                  Endpoints
                </h6>
                <div
                  class="ant-table-wrapper css-dev-only-do-not-override-7ny38l"
                >
                  <div
                    class="ant-spin-nested-loading css-dev-only-do-not-override-7ny38l"
                  >
                    <div
                      class="ant-spin-container"
                    >
                      <div
                        class="ant-table css-dev-only-do-not-override-7ny38l"
                      >
                        <div
                          class="ant-table-container"
                        >
                          <div
                            class="ant-table-content"
                          >
                            <table
                              style="table-layout: auto;"
                            >
                              <colgroup />
                              <thead
                                class="ant-table-thead"
                              >
                                <tr>
                                  <th
                                    class="ant-table-cell"
                                    scope="col"
                                  >
                                    Name
                                  </th>
                                  <th
                                    aria-label="P50  (in ms)"
                                    class="ant-table-cell ant-table-column-has-sorters"
                                    scope="col"
                                    tabindex="0"
                                  >
                                    <div
                                      class="ant-table-column-sorters"
                                    >
                                      <span
                                        class="ant-table-column-title"
                                      >
                                        P50  (in ms)
                                      </span>
                                      <span
                                        class="ant-table-column-sorter ant-table-column-sorter-full"
                                      >
                                        <span
                                          aria-hidden="true"
                                          class="ant-table-column-sorter-inner"
                                        >
                                          <span
                                            aria-label="caret-up"
                                            class="anticon anticon-caret-up ant-table-column-sorter-up"
                                            role="img"
                                          >
                                            <svg
                                              aria-hidden="true"
                                              data-icon="caret-up"
                                              fill="currentColor"
                                              focusable="false"
                                              height="1em"
                                              viewBox="0 0 1024 1024"
                                              width="1em"
                                            >
                                              <path
                                                d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                              />
                                            </svg>
                                          </span>
                                          <span
                                            aria-label="caret-down"
                                            class="anticon anticon-caret-down ant-table-column-sorter-down"
                                            role="img"
                                          >
                                            <svg
                                              aria-hidden="true"
                                              data-icon="caret-down"
                                              fill="currentColor"
                                              focusable="false"
                                              height="1em"
                                              viewBox="0 0 1024 1024"
                                              width="1em"
                                            >
                                              <path
                                                d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                              />
                                            </svg>
                                          </span>
                                        </span>
                                      </span>
                                    </div>
                                  </th>
                                  <th
                                    aria-label="P95  (in ms)"
                                    class="ant-table-cell ant-table-column-has-sorters"
                                    scope="col"
                                    tabindex="0"
                                  >
                                    <div
                                      class="ant-table-column-sorters"
                                    >
                                      <span
                                        class="ant-table-column-title"
                                      >
                                        P95  (in ms)
                                      </span>
                                      <span
                                        class="ant-table-column-sorter ant-table-column-sorter-full"
                                      >
                                        <span
                                          aria-hidden="true"
                                          class="ant-table-column-sorter-inner"
                                        >
                                          <span
                                            aria-label="caret-up"
                                            class="anticon anticon-caret-up ant-table-column-sorter-up"
                                            role="img"
                                          >
                                            <svg
                                              aria-hidden="true"
                                              data-icon="caret-up"
                                              fill="currentColor"
                                              focusable="false"
                                              height="1em"
                                              viewBox="0 0 1024 1024"
                                              width="1em"
                                            >
                                              <path
                                                d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                              />
                                            </svg>
                                          </span>
                                          <span
                                            aria-label="caret-down"
                                            class="anticon anticon-caret-down ant-table-column-sorter-down"
                                            role="img"
                                          >
                                            <svg
                                              aria-hidden="true"
                                              data-icon="caret-down"
                                              fill="currentColor"
                                              focusable="false"
                                              height="1em"
                                              viewBox="0 0 1024 1024"
                                              width="1em"
                                            >
                                              <path
                                                d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                              />
                                            </svg>
                                          </span>
                                        </span>
                                      </span>
                                    </div>
                                  </th>
                                  <th
                                    aria-label="P99  (in ms)"
                                    class="ant-table-cell ant-table-column-has-sorters"
                                    scope="col"
                                    tabindex="0"
                                  >
                                    <div
                                      class="ant-table-column-sorters"
                                    >
                                      <span
                                        class="ant-table-column-title"
                                      >
                                        P99  (in ms)
                                      </span>
                                      <span
                                        class="ant-table-column-sorter ant-table-column-sorter-full"
                                      >
                                        <span
                                          aria-hidden="true"
                                          class="ant-table-column-sorter-inner"
                                        >
                                          <span
                                            aria-label="caret-up"
                                            class="anticon anticon-caret-up ant-table-column-sorter-up"
                                            role="img"
                                          >
                                            <svg
                                              aria-hidden="true"
                                              data-icon="caret-up"
                                              fill="currentColor"
                                              focusable="false"
                                              height="1em"
                                              viewBox="0 0 1024 1024"
                                              width="1em"
                                            >
                                              <path
                                                d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                              />
                                            </svg>
                                          </span>
                                          <span
                                            aria-label="caret-down"
                                            class="anticon anticon-caret-down ant-table-column-sorter-down"
                                            role="img"
                                          >
                                            <svg
                                              aria-hidden="true"
                                              data-icon="caret-down"
                                              fill="currentColor"
                                              focusable="false"
                                              height="1em"
                                              viewBox="0 0 1024 1024"
                                              width="1em"
                                            >
                                              <path
                                                d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                              />
                                            </svg>
                                          </span>
                                        </span>
                                      </span>
                                    </div>
                                  </th>
                                  <th
                                    aria-label="Number of Calls"
                                    class="ant-table-cell ant-table-column-has-sorters"
                                    scope="col"
                                    tabindex="0"
                                  >
                                    <div
                                      class="ant-table-column-sorters"
                                    >
                                      <span
                                        class="ant-table-column-title"
                                      >
                                        Number of Calls
                                      </span>
                                      <span
                                        class="ant-table-column-sorter ant-table-column-sorter-full"
                                      >
                                        <span
                                          aria-hidden="true"
                                          class="ant-table-column-sorter-inner"
                                        >
                                          <span
                                            aria-label="caret-up"
                                            class="anticon anticon-caret-up ant-table-column-sorter-up"
                                            role="img"
                                          >
                                            <svg
                                              aria-hidden="true"
                                              data-icon="caret-up"
                                              fill="currentColor"
                                              focusable="false"
                                              height="1em"
                                              viewBox="0 0 1024 1024"
                                              width="1em"
                                            >
                                              <path
                                                d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                              />
                                            </svg>
                                          </span>
                                          <span
                                            aria-label="caret-down"
                                            class="anticon anticon-caret-down ant-table-column-sorter-down"
                                            role="img"
                                          >
                                            <svg
                                              aria-hidden="true"
                                              data-icon="caret-down"
                                              fill="currentColor"
                                              focusable="false"
                                              height="1em"
                                              viewBox="0 0 1024 1024"
                                              width="1em"
                                            >
                                              <path
                                                d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                              />
                                            </svg>
                                          </span>
                                        </span>
                                      </span>
                                    </div>
                                  </th>
                                </tr>
                              </thead>
                              <tbody
                                class="ant-table-tbody"
                              >
                                <tr
                                  class="ant-table-row ant-table-row-level-0"
                                  data-row-key="/config"
                                >
                                  <td
                                    class="ant-table-cell"
                                  >
                                    <button
                                      class="ant-btn css-dev-only-do-not-override-7ny38l ant-btn-link ant-btn-color-primary ant-btn-variant-link"
                                      type="button"
                                    >
                                      <span>
                                        /config
                                      </span>
                                      
                                    </button>
                                  </td>
                                  <td
                                    class="ant-table-cell"
                                  >
                                    0.07
                                  </td>
                                  <td
                                    class="ant-table-cell"
                                  >
                                    0.09
                                  </td>
                                  <td
                                    class="ant-table-cell"
                                  >
                                    0.10
                                  </td>
                                  <td
                                    class="ant-table-cell"
                                  >
                                    8
                                  </td>
                                </tr>
                                <tr
                                  class="ant-table-row ant-table-row-level-0"
                                  data-row-key="/dispatch"
                                >
                                  <td
                                    class="ant-table-cell"
                                  >
                                    <button
                                      class="ant-btn css-dev-only-do-not-override-7ny38l ant-btn-link ant-btn-color-primary ant-btn-variant-link"
                                      type="button"
                                    >
                                      <span>
                                        /dispatch
                                      </span>
                                      
                                    </button>
                                  </td>
                                  <td
                                    class="ant-table-cell"
                                  >
                                    765.25
                                  </td>
                                  <td
                                    class="ant-table-cell"
                                  >
                                    904.79
                                  </td>
                                  <td
                                    class="ant-table-cell"
                                  >
                                    937.20
                                  </td>
                                  <td
                                    class="ant-table-cell"
                                  >
                                    6
                                  </td>
                                </tr>
                              </tbody>
                            </table>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
